### ISSUE TITLE
/api/manifest/?sha=latest is unreliable, 404s often

### ISSUE THREAD

**Issue by tabatkins** [web-platform-tests/wpt.fyi#220]
It looks like there's a race condition in the reporting of the "latest" manifest - sometimes hitting the endpoint works great, other times it returns a 404 and reports that there's no results for a particular sha.

I suspect that while it's building a new manifest, the new sha is available in such a way that `?sha=latest` can grab it, but the actual data isn't ready yet.  We should be exposing only the latest *available* manifest data.

**Comment by tabatkins** [comment]
Hm, and even when it *does* come up in some places (currently, the latest SHA *is* coming up for me in my browser) it doesn't come up in others (the exact same URL is failing in `bikeshed update`).

**Comment by tabatkins** [comment]
What really confuses me is that the error message I get is

```
No search results found for SHA 4f7d09666f738d1db19ba49721b2db4330bf4e82
```

...which is only emitted by [/api/manifest.go#L157](https://github.com/web-platform-tests/wpt.fyi/blob/master/api/manifest.go#L157), but that particular code path looks like it's only hit when you call the function with an *actual* sha, *not* "latest" or "".

(And now it won't load in the browser, but it did load, once, for `bikeshed update`, but not it's back to not loading in the console... SHA is unchanged, so I dunno *what* is going on.)

**Comment by Hexcles** [comment]
@lukebjerring when I was investigating the 500s in my results receiver just now, I found it was caused by OOM (#221 ). On a hunch, I thought it could also cause manifest API to fail. And after examining the logs, the flakiness of this API is indeed caused by OOM. Example:

```
GET /api/manifest?path=/css/css-flexbox/auto-margins-001.html
Exceeded soft private memory limit of 128 MB with 136 MB after servicing 1553 requests total
After handling this request, the process that handled this request was found to be using too much memory and was terminated. This is likely to cause a new process to be used for the next request to your application. If you see this message frequently, you may have a memory leak in your application.
```

I don't know how much memory the manifest API needs. Instead of spending time on trial and error, shall we just go for the largest instance (1GB memory)? After monitoring the memory usage for a while, we can scale it down if we don't need that much memory.

**Comment by Hexcles** [comment]
BTW, inside the API, "latest" gets translated to a full hash which is (or close to, due to caching) the latest commit on WPT master. This is where the SHA comes from. @lukebjerring can tell you more.

To clarify, I'm not sure OOM is the root cause here. It's definitely an issue, but there might be others.

**Comment by tabatkins** [comment]
Ah, when I was tracing it by hand, it didn't *look* like it called back into itself with a known SHA, but I could definitely have missed it.

An OOM would explain the intermittent issues, tho.  I *can* reliably get to the `.json.gz` file named in <https://api.github.com/repos/web-platform-tests/wpt/releases/latest>, and the API endpoint seems to just unpack that file to raw JSON; can we instead just *host* the raw JSON directly so we don't have to pass thru an in-memory process to unpack and serve it?

**Comment by lukebjerring** [comment]
Internally, the implementation caches the fetched manifests under their SHA. For `latest`, it actually caches the latest manifest's SHA, so loads that from cache then loads the actual manifest from cache by that SHA. It's not an HTTP redirect.

Let us know if #221 improves things here.

**Comment by Hexcles** [comment]
@tabatkins actually, #221 hasn't been deployed to prod. Could you try `staging.wpt.fyi/...` instead?

**Comment by Hexcles** [comment]
The last occurance of OOM in staging was 2018-06-04 18:15:24.763 EDT; there's been no OOM since my fix (#221) was deployed to `staging.wpt.fyi`. Closing this issue.

Note that we haven't deployed to prod yet, but we will soon.
